### PR TITLE
[Yaml] Feature #48920  Allow milliseconds and microseconds in dates

### PR DESCRIPTION
**Pull request by dustinwilson**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Feature #48920
| License       | MIT
| Doc PR        | forthcoming

Allows Yaml to parse dates with milliseconds or microseconds and to dump them as well.

### CODE REVIEW

File: CHANGELOG-6.2.md [hunk 1]
```diff
@@ -331,4 +331,4 @@ To get the diff between two versions, go to https://github.com/symfony/symfony/c
  * feature #46430 [Routing] Add `Requirement::POSITIVE_INT` for common ids and pagination (HeahDude)
  * feature #46279 [DependencyInjection] Optimize autowiring logic by telling it about excluded symbols (nicolas-grekas)
  * feature #46452 [DependencyInjection] Add Hydrator::hydrate() and preserve PHP references when using it (nicolas-grekas)
-
+ * feature #48920 [Yaml] Allow milliseconds and microseconds in dates (dustinwilson)
```
Comment (GromNaN) [inline CHANGELOG-6.2.md]: The changelog is updated automatically for new releases. You can revert this change.

File: src/Symfony/Component/Yaml/Tests/InlineTest.php [hunk 2]
```diff
@@ -563,9 +563,10 @@ public function getTestsForDump()
     /**
      * @dataProvider getTimestampTests
      */
-    public function testParseTimestampAsUnixTimestampByDefault(string $yaml, int $year, int $month, int $day, int $hour, int $minute, int $second)
+    public function testParseTimestampAsUnixTimestampByDefault(string $yaml, int $year, int $month, int $day, int $hour, int $minute, int $second, int $microsecond)
     {
-        $this->assertSame(gmmktime($hour, $minute, $second, $month, $day, $year), Inline::parse($yaml));
+        $expectedDate = (new \DateTimeImmutable($yaml))->format('U');
+        $this->assertSame(($microsecond) ? (float) "$expectedDate.$microsecond" : (int) $expectedDate, Inline::parse($yaml));
```
Comment (GromNaN) [inline src/Symfony/Component/Yaml/Tests/InlineTest.php]: Parenthesis are not needed.

```suggestion
        $this->assertSame($microsecond ? (float) "$expectedDate.$microsecond" : (int) $expectedDate, Inline::parse($yaml));
```

File: src/Symfony/Component/Yaml/Inline.php [hunk 3]
```diff
@@ -712,15 +721,20 @@ private static function evaluateScalar(string $scalar, int $flags, array &$refer
                             return $time;
                         }
 
-                        try {
-                            if (false !== $scalar = $time->getTimestamp()) {
-                                return $scalar;
+                        $length = \strlen(rtrim($time->format('u'), '0'));
+                        if (0 === $length) {
+                            try {
+                                if (false !== $scalar = $time->getTimestamp()) {
```
Comment (GromNaN) [inline src/Symfony/Component/Yaml/Inline.php:727]: Do you know why there is 2 ways of generating a timestamp with a possible `false` or `TypeError`?

**Review by GromNaN — CHANGES_REQUESTED** [review]
The implementation looks good.

I have some comments to improve the code.

File: src/Symfony/Component/Yaml/Inline.php (hunk 3 above)
Comment (dustinwilson) [reply src/Symfony/Component/Yaml/Inline.php:727]: No. It's in the existing code. There's no comment explaining the reason, but tests pass without it.

Comment (dustinwilson) [reply src/Symfony/Component/Yaml/Inline.php:727]: Found it. [2353f19cc2c99c34cd1f126bd95f042cbe458824](https://github.com/symfony/symfony/commit/2353f19cc2c99c34cd1f126bd95f042cbe458824)

**Review by GromNaN — APPROVED** [review]

**Comment by dustinwilson** [comment]
@GromNaN The documentation mentions I need to squash before the pull request will be accepted do I still need to do this?

**Comment by stof** [comment]
@dustinwilson if you allow maintainers to edit your PR (which is the case for this PR), our tooling can manage the squashing automatically.

**Comment by dustinwilson** [comment]
@stof Okay. Thanks!

File: CHANGELOG-6.2.md [hunk 4]
```diff
@@ -331,4 +331,3 @@ To get the diff between two versions, go to https://github.com/symfony/symfony/c
  * feature #46430 [Routing] Add `Requirement::POSITIVE_INT` for common ids and pagination (HeahDude)
  * feature #46279 [DependencyInjection] Optimize autowiring logic by telling it about excluded symbols (nicolas-grekas)
  * feature #46452 [DependencyInjection] Add Hydrator::hydrate() and preserve PHP references when using it (nicolas-grekas)
-
```
Comment (fabpot) [inline CHANGELOG-6.2.md]: Should be reverted

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @dustinwilson.
